subsys/net/lib/fota_download/src/fota_download.c

```diff
@@ -7,22 +7,250 @@
 #include <stdint.h>
 #include <zephyr.h>
 #include <flash.h>
-#include <net/download_client.h>
 #include <dfu/mcuboot.h>
 #include <dfu/flash_img.h>
 #include <pm_config.h>
 #include <logging/log.h>
+#include <net/download_client.h>
 #include <net/fota_download.h>
+#include <nrf_socket.h>
+#include <net/socket.h>
 
 LOG_MODULE_REGISTER(fota_download, CONFIG_FOTA_DOWNLOAD_LOG_LEVEL);
 
+static struct	download_client		 dfu;
+static struct	flash_img_context	 flash_img;
 static		fota_download_callback_t callback;
-static struct	flash_img_context flash_img;
-static struct	download_client dfu;
+
+#ifdef CONFIG_FOTA_DOWNLOAD_MODEM_FW_UPDATE_SUPPORT
+static int			modem_dfu_socket_fd;
+static struct	k_sem		modem_fw_delete_sem;
+static struct	k_delayed_work  modem_fw_delete_dwork;
+static struct k_work_q modem_fw_delete_offload_work_q;
+static K_THREAD_STACK_DEFINE(modem_fw_delete_offload_work_q_stack,
+			     CONFIG_FOTA_DOWNLOAD_OFFLOAD_WORKQUEUE_STACK_SIZE);
+
+static int modem_fw_update(void)
+{
+	int rc;
+
+	LOG_INF("Scheduling modem firmware update at next boot");
+	rc = nrf_setsockopt(modem_dfu_socket_fd, NRF_SOL_DFU, NRF_SO_DFU_APPLY,
+			    NULL, 0);
+	if (rc < 0) {
+		LOG_INF("Failed to schedule update, errno %d", errno);
+	}
+
+	return 0;
+}
+
+static int modem_fw_delete(void)
+{
+	int rc;
+
+	LOG_INF("Deleting firmware image, please wait...\n");
+	rc = nrf_setsockopt(modem_dfu_socket_fd, NRF_SOL_DFU,
+			    NRF_SO_DFU_BACKUP_DELETE, NULL, 0);
+	if (rc < 0) {
+		LOG_ERR("Failed to delete backup, errno %d\n", errno);
+		return 0;
+	}
+
+	k_delayed_work_submit_to_queue(&modem_fw_delete_offload_work_q,
+				       &modem_fw_delete_dwork, K_SECONDS(1));
+
+	return 0;
+}
+
+/**@brief Initialize DFU socket. */
+static int modem_dfu_socket_init(void)
+{
+	int rc;
+	nrf_socklen_t len;
+	nrf_dfu_fw_version_t version;
+	nrf_dfu_resources_t resource;
+	nrf_dfu_fw_offset_t off;
+
+	/* Create a socket for firmware update */
+	modem_dfu_socket_fd = nrf_socket(NRF_AF_LOCAL, NRF_SOCK_STREAM,
+				     NRF_PROTO_DFU);
+	if (modem_dfu_socket_fd < 0) {
+		LOG_ERR("Failed to open Modem DFU socket.");
+		return -1;
```

Comment:
Here and below - you could return the actual error code (`-errno`) instead of hiding it with -1. Especially that we log it after `modem_dfu_socket_init` exits.

Reply:
fix